polar-core/src/vm.rs

```diff
@@ -2289,41 +2289,46 @@ impl PolarVirtualMachine {
-                    // This is done here for safety to avoid a bug where `answer` is unbound by
-                    // `IsSubspecializer` and the `Unify` Goal just assigns it to `true` instead
-                    // of checking that is is equal to `true`.
-                    self.bind(&answer, Term::new_temporary(Value::Boolean(false)));
-
-                    return self.append_goals(vec![
-                        Goal::IsSubspecializer {
-                            answer: answer.clone(),
-                            left: left_spec.clone(),
-                            right: right_spec.clone(),
-                            arg: arg.clone(),
-                        },
-                        Goal::Unify {
-                            left: Term::new_temporary(Value::Variable(answer)),
-                            right: Term::new_temporary(Value::Boolean(true)),
-                        },
-                    ]);
+            match (&left_param.specializer, &right_param.specializer) {
+                (Some(left_spec), Some(right_spec)) => {
+                    // If you find two non-equal specializers, that comparison determines the relative
+                    // specificity of the two rules completely. As soon as you have two specializers
+                    // that aren't the same and you can compare them and ask which one is more specific
+                    // to the relevant argument, you're done.
+                    if left_spec != right_spec {
+                        let answer = self.kb.read().unwrap().gensym("is_subspecializer");
+                        // Bind answer to false as a starting point in case is subspecializer doesn't
+                        // bind any result.
+                        // This is done here for safety to avoid a bug where `answer` is unbound by
+                        // `IsSubspecializer` and the `Unify` Goal just assigns it to `true` instead
+                        // of checking that is is equal to `true`.
+                        self.bind(&answer, Term::new_temporary(Value::Boolean(false)));
+
+                        return self.append_goals(vec![
+                            Goal::IsSubspecializer {
+                                answer: answer.clone(),
+                                left: left_spec.clone(),
+                                right: right_spec.clone(),
+                                arg: arg.clone(),
+                            },
+                            Goal::Unify {
+                                left: Term::new_temporary(Value::Variable(answer)),
+                                right: Term::new_temporary(Value::Boolean(true)),
+                            },
+                        ]);
+                    }
                 }
+                // If the left rule has no specializer and the right does, it is NOT more specific,
+                // so we fall through to Backtrack (fail)
+                (None, Some(_)) => (),
```

Comment:
This needs to either break from the loop or return backtrack immediately (probably preferable).  Otherwise, a multiple argument rule will not work properly (and we should add a test for that).

Reply:
Right. So the behavior is that the first specializer determines ordering.

Reply:
This is in a for loop so this will continue to the next specializer, right, not fail? I'm thinking of a call like:

```
left: f(a, b: B), right: f(a: A, b))
```

This would succeed because the first iteration would continue, and the second return Ok.  Instead, we want the first iteration to immediately backtrack.